### ISSUE TITLE
TriMesh magics not effective with datashader

### ISSUE THREAD

**Issue by kcpevey** [ioam/holoviews#2651]
It appears that cell magics (and global) do not have an effect when visualizing `TriMesh.edgepaths` via datashader.   
  
```
%%opts TriMesh (node_size=0 edge_line_width=0.1, edge_line_color='blue')
tri_mesh = hv.TriMesh((tris, points))
```
  
Just visualizing via `tri_mesh` call renders the wireframe and picks up the TriMesh opts. But calling `datashade(tri_mesh.edgepaths)` will ignore all my TriMesh opts.

**Comment by jlstevens** [comment]
The options you supply to ``TriMesh`` *without* using datashader are processed and passed onto Bokeh. When using ``datashade``, it is datashader that is in charge of rendering the graph which doesn't support the same set of options as bokeh.

Datashader really has a very different rendering model from Bokeh specified in a rather different way (e.g using concepts such as aggregators). This model is suited to rendering very large data without overplotting but it doesn't offer the same sort of options as when rendering with Bokeh. 

In other words, this is the expected behavior although we should now use the datashader options that are available to get as close to the results that you want as we can. Perhaps @jbednar and @philippjfr have some suggestions? At the very least, we should be able to set what is effectively the edge color, and maybe the width...

**Comment by jbednar** [comment]
You can do `hv.help(datashade)` to see the arguments available for controlling `datashade`. Some of these are comparable to what is available for Bokeh, and for those I've proposed that we make some sort of automatic mapping from the plot option into a datashader option (see https://github.com/ioam/holoviews/issues/2637). However, only one of the three options you list above currently has any direct mapping into what datashader supports, so even if we did implement such mapping, it wouldn't fully solve your immediate issue.

I'd also be surprised if those options even applied to `tri_mesh.edgepaths` without `datashade()`, given that EdgePaths is a separate element, but in that case it would suffice to change TriMesh to EdgePaths in the options magic.

In any case, for the moment at least, we have to step back and think about what you are trying to achieve.  I.e., is there anything available from datashader that can help achieve it?  Assuming you replaced TriMesh with EdgePaths, it looks to me like without `datashade()` you'd get a wireframe plot with blue lines and a certain width.  To set the color with`datashade()`, you could pass `cmap=['blue']`; datashader will usually use a wide range of colors depending on overlaps, and using a single-color list like that we can force it to act like Bokeh.  Then to make the lines thicker, you can use `spread(datashade(...), max_px=2)` to spread each active pixel by 2 in each direction.  Datashader's spreading is a bit weird, acting on the pixel values after drawing rather than actually rendering a thick line, but it can give a similar result of making the line thicker.  You'll need to import `spread` first, leading to this code (untested since I don't have your example):

```
from holoviews.operation.datashader import datashade, spread
spread(datashade(tri_mesh.edgepaths, cmap=['blue']), max_px=2)
```

(Where you have to adjust the value 2 until it looks like the thickness you want).  Sorry this is awkward; Bokeh and Datashader do really have quite different models of how to render data, and it's not simple to translate between them.

**Comment by kcpevey** [comment]
I think I understand, at least enough to achieve my goals. On a very basic level, I'm just wanting to display a mesh in its two basic forms: wire mesh and color contouring based on elevation.   
  
Instead of using datashade for both of those operations, I'm switching to plain trimesh for the wire frame/edgepaths and using datashade for the aggregated color contours and using overlays to achieve both in the same drawing.   
    
For example, now I have: `datashade(tri_mesh, aggregator=ds.mean('z')) * tri_mesh.edgepaths` which works great for me.   
  
Just FYI, I was led astray by this example notebook (cell 3) https://github.com/pyviz/EarthSim/blob/master/examples/user_guide/Visualizing_Meshes.ipynb
  
As far as I'm concerned this can be closed :)

**Comment by kcpevey** [comment]
Also, a note for myself if no one else... the suggested method of `datashade(trimesh.edgepaths, cmap=['black'])` achieves the same purpose while allowing for big data to be handled. `trimesh.edgepaths` alone works, but is much less effective at handling medium/large meshes (40k+ nodes).
